### ISSUE TITLE
script exits with "done", Folium/Sudachi still crashing

### ISSUE THREAD

**Issue by Petee464** [fritzlb/iOS17-JIT-WIN#34]
Hi, I run the script as told to, and it says "done." at the end, however, Sudachi is still crashing when I try to open a game. Any idea why?
```
Starting debug server...
(b"\r\nFollow the following connections steps from LLDB:\r\n\r\n(lldb) platform select remote-ios\r\n(lldb) target create /path/to/local/application.app\r\n(lldb) script lldb.target.module[0].SetPlatformFileSpec(lldb.SBFileSpec('/private/var/containers/Bundle/Application/<APP-UUID>/application.app'))\r\n(lldb) process connect connect://[fdcd:7c1e:7feb::1]:60134   <-- ACTUAL CONNECTION DETAILS!\r\n(lldb) process launch\r\n\r\n", b'')
Started debug server with connection details: [fdcd:7c1e:7feb::1]:60134
Run debugging commands...
This might take a few minutes.
(lldb) command source -s 0 cmdfile.txt
Executing commands in 'C:\Users\POMAHZ~1\Desktop\IOS17-~1\cmdfile.txt'.
(lldb) gdb-remote [fdcd:7c1e:7feb::1]:60134
(lldb) settings set target.memory-module-load-level minimal
(lldb) attach -p 6716
Process 6716 stopped
* thread #1, stop reason = signal SIGSTOP
    frame #0: 0x00000001e89e9af8 libsystem_kernel.dylib
->  0x1e89e9af8: ret
    0x1e89e9afc: mov    x16, #-0x30               ; =-48
    0x1e89e9b00: svc    #0x80
    0x1e89e9b04: ret
Executable module set to "\private\var\containers\Bundle\Application\C6B31A47-EE9B-400B-9607-0FABE4BDB50C\App.app\Folium".
(lldb) detach
Process 6716 detached
(lldb) (lldb)  quit
 Could not find implementation lookup function "class_getMethodImplementation" step in through ObjC method dispatch will not work.
warning: libobjc.A.dylib is being read from process memory. This indicates that LLDB could not find the on-disk shared cache for this device. This will likely reduce debugging performance.

(lldb) warning: Architecture changed from arm64e-apple-ios to arm64-apple-ios.

done.
DEBUG mode specified. This script won't quit automatically. Press CTRL + C to stop it.
```

**Comment by fritzlb** [comment]
I didn’t test Sudachi bc I do not own any compatible games. Any chance you could test an app that’s known to work?
Also, in case you‘re on an iPhone 15 (pro), most emulators apparently require adjustments. In case another app works, you‘re probably better off contacting the developer of sudachi. If nothing works I‘d try to troubleshoot that first, if you get stuck I‘ll take a look at it :)

**Comment by Petee464** [comment]
Hi, it might be possible that sudachi also requires extended ram entitlements, I'll check how to turn that on, and regarding trying another app, I'm not aware of any that would entirely cease to function without JIT, so what's a foolproof method of checking whether it's on?

**Comment by fritzlb** [comment]
For example starting a VM in UTM or starting minecraft (preferably an older version like 1.18 or older) (or any other supported java program, like a mod installer) in pojavlauncher is a good indicator :)

Also, make sure to sideload the app using a real development certificate, esign doesn't work. AltStore and Sideloadly Apple ID sideloading are known to work.

**Comment by pear-studios** [comment]
I’ve been doing some testing on my own of this situation, and I think I have figured out the problem.

Before that, I just wanted to let anyone know that Sudachi is a very new and very buggy software, so crashes are bound to happen.

As fritzlb mentioned, trying to use the software to activate other apps with JIT would be a good idea to see where the problem is coming from. Good examples would be UTM, DolphiniOS, and PojavLauncher. We could then figure out if it’s a software problem or not.

As you probably know, Sudachi requires a powerful system to run, especially since they’re switch games being emulated. To be precise, around 6 GB of RAM is required to run anything, which isn’t very common among iOS devices.

I have tried running Sudachi on a 10th Gen iPad, which to no surprise, did not run because it had 4 GB of RAM. I do have an iPhone 14 that has 6 GB of RAM, but it’s running iOS 17 (I haven’t tried this program out, will soon).

I will do more testing on my iPhone, but it would be really helpful if you could comment the system that you’re running on to see if it’s a hardware problem, like you mentioned.

**Comment by pear-studios** [comment]
Also, fritzlb, the end of your latest comment was a little confusing to me. Can you elaborate and the E-signing stuff and development certificates?

**Comment by Petee464** [comment]
hi pear-studios;
I'm on a 15 pro max so I doubt hardware would be an issue

**Comment by pear-studios** [comment]
I see. The creator did use an iPhone 15 Pro Max, although he was running on iOS 17.2. What iOS are you using?

**Comment by fritzlb** [comment]
> Also, fritzlb, the end of your latest comment was a little confusing to me. Can you elaborate and the E-signing stuff and development certificates?

Yeah, of course. This might get a little long, as there‘s a bunch of theory behind it. Also everything in the following text might not be 100% accurate because stuff gets complex pretty fast and I definitely don’t know or understand everything, but here are the things I read somewhere:

First of all, let’s start with the device booting process. Apple wants their devices to be secure (someone could even say they‘d like to control everything about it), so they use a secure boot method. That basically means somewhere etched in the hardware is a tiny program which gets started first and checks the signature of the installed system. If the signature doesn’t match that means someone except Apple tampered with the system and it won’t start. Every component of iOS is built on top of that and checks the signature of the next component that gets loaded, if Apple didn’t sign it, it‘ll refuse to work. 
While android uses similar technologies, they do allow apps to be installed from anywhere (with according settings ofc). 
Apple does things differently. They check the signature of the app every time before it gets launched. That means: no signature from Apple -> app doesn’t work. If you install an app from the App Store, Apple will go ahead, sign it for you so that the App Store can install it. 
Also, iOS devices have a bunch of special hardware components designed with this system in mind. For example RAM can either be set as „writable“ or as „executable“. Apps will get placed into executable sections, while data (such as texts, images, …) are in writable sections (bc you might want to edit a document or photo).

For apps requiring JIT this is a problem, because JIT (just in time) compilers begin by interpreting the code which is pretty slow and then basically take a look at what they‘ve done and try to optimize it. For example, if you program something like a loop that just adds 1 to a number, but 1000 times so that the number gets to 1000, an interpreter would actually go ahead and do 1000 calculations while a JIT compiler would „see“ what the code is doing and instead of adding 1000 ones do one operation and add 1000.
For that, they do have to write to ram sections they execute afterwards. Which is impossible on iOS (without a jailbreak). 

Unless your app is specifically compiled as a developer app. See, for developers it might be pretty hard to see why a specific bug is happening. For that purpose there exist debuggers. A debugger is a program that can watch and influence another program while it’s running. A developer can use such a debugger for example to stop the app while something happens, see variables etc. To do this, there must be a possibility to write to ram and execute the contents. So that’s not possible on iOS?
No, actually it is. If an app is installed using a developer certificate (from a free or paid developer account), those special privileges are being given to the app. But they‘re not being taken away once the debugger disconnects, only when the app is completely stopped. 

Apps requiring JIT use exactly this trick. That means it’s absolutely necessary to install them using a developer certificate. To enable JIT then it’s enough to open the app and attach a debugger for a few seconds, that’s exactly what my script does. 

AltStore does use this method, the ipa file of the app gets signed from Apple with the help of your Apple ID credentials and depending on wether you‘re using a free or paid account the certificate lasts for a year or one week. Paid developer account certificates do have other capabilities as well, but let’s not get into that. 

Sideloadly usually uses the same method as AltStore for sideloading, so there’s no difference if you‘re using it with your Apple ID. Sideloadly can be used with a certificate tho afaik, in that case it depends on the certificate if the app works with JIT. 

Other sideloading tools, like esign, especially the ones not requiring a computer (except sidestore, they use vpn tricks) oftentimes use enterprise certificates. Those certificates can be used by companies to distribute internal apps that should not be present in the App Store for example. But those certificates aren’t development certificates, so ios blocks debugging those apps which blocks JIT. 

In my opinion it’s kinda sad Apple blocks JIT in every possible case, the m1/m2 iPads for example are extremely powerful but severely held back by iOS. While technically they can run a a VM, the Java version of Minecraft, and all kinds of cool emulators, Apple doesn’t let you. 
Kinda similar to what we see in the EU with sideloading. Apple is holding back the iPhone (and iPads) with the AppStore. But let’s not get into that.
